controller/canary.go

```diff
@@ -83,6 +83,12 @@ func (c *Controller) rolloutCanary(rollout *v1alpha1.Rollout, rsList []*appsv1.R
 		logCtx.Infof("Not finished reconciling Canary Pause")
 		return c.syncRolloutStatusCanary(oldRSs, newRS, stableRS, rollout)
 	}
+	if conditions.RolloutComplete(rollout, &rollout.Status) {
+		logCtx.Info("Cleaning up old replicasets")
+		if err := c.cleanupRollouts(oldRSs, rollout); err != nil {
+			return err
+		}
+	}
```

Comment:
This code occurs too late, and we will never hit this logic if, for example, the rollout canary strategy has pauses in it. It's possible to deploy many versions of a rollout with a pause step, and then reach a greater number of replicasets than what is controlled in `spec.revisionHistoryLimit`. The cleanup of rollouts needs to happen every reconciliation, and should be a no-op if there is nothing to clean up.